go/pkg/rexec/rexec.go

```diff
@@ -332,35 +341,31 @@ func (ec *Context) ExecuteRemotely() {
 	ec.Metadata.RealBytesUploaded = bytesMoved
 	log.V(1).Infof("%s %s> Executing remotely...\n%s", cmdID, executionID, strings.Join(ec.cmd.Args, " "))
 	ec.Metadata.EventTimes[command.EventExecuteRemotely] = &command.TimeInterval{From: time.Now()}
-	eg, ctx := errgroup.WithContext(ec.ctx)
-	var streamOut, streamErr sync.Once
-	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ctx, &repb.ExecuteRequest{
+	// These variables are owned by the progress callback (which is async but not conrrent) until the execution returns.
+	var streaming bool
+	var nOutStreamed, nErrStreamed int64
+	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ec.ctx, &repb.ExecuteRequest{
 		InstanceName:    ec.client.GrpcClient.InstanceName,
 		SkipCacheLookup: !ec.opt.AcceptCached || ec.opt.DoNotCache,
 		ActionDigest:    ec.Metadata.ActionDigest.ToProto(),
 	}, func(md *repb.ExecuteOperationMetadata) {
-		if !ec.opt.StreamOutErr {
+		if streaming || !ec.opt.StreamOutErr {
 			return
 		}
+		streaming = true
```

Comment:
This assumes that if the stdout/stderr stream names are provided, they are always provided on the first callback. I think RBE does always do this, but is this a requirement of the API? Or would it be technically valid to provide progress callbacks for a while before you end up sending a callback that includes the stream names?

Reply:
Nothing is specified in [the API](https://github.com/bazelbuild/remote-apis/blob/068363a3625e166056c155f6441cfb35ca8dfbf2/build/bazel/remote/execution/v2/remote_execution.proto#L118), which means the server may return either, neither or both, and not necessarily in the same or first call.

Good call!